### PR TITLE
8310234: Refactor Locale tests to use JUnit

### PR DESCRIPTION
**Pull request by justin-curtis-lu**
Please review this PR as apart of [JDK-8307843](https://bugs.openjdk.org/browse/JDK-8307843) which refactors some tests in Locale to use JUnit. Other cleanup and small changes are included as well. More refactoring in Locale tests will be done in separate PRs.

If the test had a bugNNNNN.java name, it was also renamed to something more [descriptive](https://openjdk.org/jtreg/faq.html#how-should-i-name-a-test).

Below is a list of all the changes,

- Refactor Bug4316602.java as LocaleConstructors.java
- Refactor Bug4210525.java as CaseCheckVariant.java
- Refactor bug6277243.java as RootLocale.java
- Refactor bug6312358.java as GetInstanceCheck.java
- Refactor Bug8154797.java as CompareProviderFormats.java
- Refactor Bug8004240.java as GetAdapterPreference.java
- Refactor bug4122700.java into AvailableLocalesTest.java (and combined with StreamAvailableLocales.java)

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8310234](https://bugs.openjdk.org/browse/JDK-8310234): Refactor Locale tests to use JUnit (**Sub-task** - P4)

### Reviewers
 * [Naoto Sato](https://openjdk.org/census#naoto) (@naotoj - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/14609/head:pull/14609` \
`$ git checkout pull/14609`

Update a local copy of the PR: \
`$ git checkout pull/14609` \
`$ git pull https://git.openjdk.org/jdk.git pull/14609/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 14609`

View PR using the GUI difftool: \
`$ git pr show -t 14609`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/14609.diff">https://git.openjdk.org/jdk/pull/14609.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/14609#issuecomment-1601884628)

### CODE REVIEW

File: test/jdk/java/util/Locale/LocaleConstructors.java [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+public class LocaleConstructors {
+
+    static final String LANG = "en";
+    static final String COUNTRY = "US";
+    static final String VAR = "socal";
+
+    @Test
+    public void langTest() {
+        Locale aLocale = Locale.of(LANG);
+        Locale otherLocale = new Locale(LANG);
```
Comment (justin-curtis-lu) [inline test/jdk/java/util/Locale/LocaleConstructors.java:51]: Is there a preference on testing code that is marked as `@deprecated` in the JDK? This test originally validated the parameters of the deprecated Locale constructor but was changed in 2022 with the Locale.of() method. However, the test was originally meant to test the constructor so I re-inserted it here.

Comment (LanceAndersen) [reply test/jdk/java/util/Locale/LocaleConstructors.java:51]: If the functionality still exists we should keep the existing test(s)

File: test/jdk/java/util/Locale/LocaleConstructors.java [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+        assertEquals(aLocale.toString(), LANG);
+        assertEquals(otherLocale.toString(), LANG);
+    }
+
+    @Test
```
Comment (LanceAndersen) [inline test/jdk/java/util/Locale/LocaleConstructors.java:56]: Minor, but please consider a simple comment describing the tests intent.  Just makes it easier for future developers...

File: test/jdk/java/util/Locale/LocaleConstructors.java (hunk 1 above)
Comment (justin-curtis-lu) [reply test/jdk/java/util/Locale/LocaleConstructors.java:52]: Got it, will leave it in then.

File: test/jdk/java/util/Locale/LocaleConstructors.java (hunk 2 above)
Comment (justin-curtis-lu) [reply test/jdk/java/util/Locale/LocaleConstructors.java:58]: Had originally only included a blanket comment for the test class. Updated the file with a comment for each test method to clarify intent.

File: test/jdk/java/util/Locale/GetInstanceCheck.java [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+     */
+    @ParameterizedTest
+    @MethodSource("failingArguments")
+    public void throwNPETest(String language, String country, String variant) {
+        try {
+            getInstanceMethod.invoke(null, language, country, variant);
+            fail("Should NPE with any argument set to null");
+        } catch (InvocationTargetException | IllegalAccessException exc) {
+            Throwable cause = exc.getCause();
+            if (cause instanceof NullPointerException) {
```
Comment (naotoj) [inline test/jdk/java/util/Locale/GetInstanceCheck.java:99]: Same as above

File: test/jdk/java/util/Locale/GetInstanceCheck.java [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+     * an NPE if no arguments are null.
+     */
+    @ParameterizedTest
+    @MethodSource("passingArguments")
+    public void noNPETest(String language, String country, String variant) {
+        try {
+            getInstanceMethod.invoke(null, language, country, variant);
+        } catch (InvocationTargetException | IllegalAccessException exc) {
+            Throwable cause = exc.getCause();
+            if (exc.getCause() instanceof NullPointerException) {
```
Comment (naotoj) [inline test/jdk/java/util/Locale/GetInstanceCheck.java:79]: Should this condition only be meaningful for `InvocationTargetException`?

File: test/jdk/java/util/Locale/AvailableLocalesTest.java [hunk 5]
```diff
@@ -20,30 +20,45 @@
...
 import org.junit.jupiter.params.provider.Arguments;
 
-public class StreamAvailableLocales {
+import static org.junit.jupiter.api.Assertions.assertNotEquals;
+
+public class AvailableLocalesTest {
+
+    /**
+     * Test that Locale.getAvailableLocales() is non-empty.
+     * Print out the locales.
```
[the remark this answers is not in the document]
Comment (justin-curtis-lu) [reply test/jdk/java/util/Locale/AvailableLocalesTest.java:46]: Unless I am mistaken, both bug IDs, 4122700 8282319 are in the combined file.

File: test/jdk/java/util/Locale/AvailableLocalesTest.java [hunk 6]
```diff
@@ -20,30 +20,45 @@
...
+        printLocales(systemLocales);
+    }
 
     /**
      * Test to validate that the methods: Locale.getAvailableLocales()
      * and Locale.availableLocales() contain the same underlying elements
      */
     @Test
-    public void testStreamEqualsArray() {
+    public void StreamEqualsArrayTest() {
```
[the remark this answers is not in the document]
Comment (justin-curtis-lu) [reply test/jdk/java/util/Locale/AvailableLocalesTest.java]: Missed that, thank you.

File: test/jdk/java/util/Locale/CaseCheckVariant.java [hunk 7]
```diff
@@ -20,30 +20,43 @@
...
+    static final String LANG = "en";
+    static final String COUNTRY = "US";
+
+    /**
+     * When a locale is created with a given variant, ensure
+     * that the variant is not case normalized.
+     */
+    @ParameterizedTest
+    @MethodSource("variants")
+    public void variantNormalizationTest(String variant) {
```
[the remark this answers is not in the document]
Comment (justin-curtis-lu) [reply test/jdk/java/util/Locale/CaseCheckVariant.java]: That's a better name, fixed.

File: test/jdk/java/util/Locale/GetInstanceCheck.java (hunk 4 above)
Comment (justin-curtis-lu) [reply test/jdk/java/util/Locale/GetInstanceCheck.java:81]: Yes you're right, that's a good point. Reserved the catch clause to handle `InvocationTargetException` and moved `IllegalAccessException` to the method signature.

File: test/jdk/java/util/Locale/AvailableLocalesTest.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (naotoj) [reply test/jdk/java/util/Locale/AvailableLocalesTest.java:46]: Yes. I wanted to make it clear that this `nonEmptyLocalesTest` was for the issue 4122700.

[the remark this answers is not in the document]
Comment (justin-curtis-lu) [reply test/jdk/java/util/Locale/AvailableLocalesTest.java]: Sorry, I totally misread you said 'method'. Fixed `nonEmptyLocalesTest` to include the bug ID.

**Review by naotoj — APPROVED** [review]
LGTM

**Pull request closed without merging**